File: android/src/main/java/com/iconshot/detonator/element/TextAreaElement.java
package com.iconshot.detonator.element;

import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.util.TypedValue;
import android.widget.EditText;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.helpers.ColorHelper;
import com.iconshot.detonator.helpers.CompareHelper;
import com.iconshot.detonator.helpers.ContextHelper;
import com.iconshot.detonator.helpers.PixelHelper;

public class TextAreaElement extends Element<EditText, TextAreaElement.Attributes> {
    private int defaultColor;
    private int defaultPlaceholderColor;

    public TextAreaElement(Detonator detonator) {
        super(detonator);
    }

    @Override
    public Class<Attributes> getAttributesClass() {
        return Attributes.class;
    }

    @Override
    protected EditText createView() {
        EditText view = new EditText(ContextHelper.context);

        defaultColor = view.getCurrentTextColor();

        TextWatcher textWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                OnChangeData data = new OnChangeData();

                data.value = s.toString();

                detonator.emitHandler("onChange", edge.id, data);
            }

            @Override
            public void afterTextChanged(Editable s) {}
        };

        view.addTextChangedListener(textWatcher);

        return view;
    }

    @Override
    protected void patchView() {
        String placeholder = attributes.placeholder;

        String currentPlaceholder = currentAttributes != null ? currentAttributes.placeholder : null;

        boolean patchPlaceholder = forcePatch || !CompareHelper.compareObjects(placeholder, currentPlaceholder);

        if (patchPlaceholder) {
            view.setHint(placeholder != null ? placeholder : "");
        }

        Object placeholderColor = attributes.placeholderColor;

        Object currentPlaceholderColor = currentAttributes != null ? currentAttributes.placeholderColor : null;

        boolean patchPlaceholderColor = forcePatch || !CompareHelper.compareColors(placeholderColor, currentPlaceholderColor);

        if (patchPlaceholderColor) {
            Integer tmpColor = ColorHelper.parseColor(placeholderColor);

            view.setHintTextColor(tmpColor != null ? tmpColor : defaultPlaceholderColor);
        }

        String value = attributes.value;

        boolean patchValue = forcePatch;

        if (patchValue) {
            view.setText(value != null ? value : "");
        }

        String autoCapitalize = attributes.autoCapitalize;

        String currentAutoCapitalize = currentAttributes != null ? currentAttributes.autoCapitalize : null;

        boolean patchAutoCapitalize = forcePatch || !CompareHelper.compareObjects(autoCapitalize, currentAutoCapitalize);

        if (patchAutoCapitalize) {
            patchInputType(autoCapitalize);
        }
    }

    private void patchInputType( String autoCapitalize) {
        int value = InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_FLAG_MULTI_LINE;

        String tmpAutoCapitalize = autoCapitalize != null ? autoCapitalize : "sentences";

        switch (tmpAutoCapitalize) {
            case "characters": {
                value |= InputType.TYPE_TEXT_FLAG_CAP_CHARACTERS;

                break;
            }

            case "words": {
                value |= InputType.TYPE_TEXT_FLAG_CAP_WORDS;

                break;
            }

            case "sentences": {
                value |= InputType.TYPE_TEXT_FLAG_CAP_SENTENCES;

                break;
            }

            case "none": {
                value |= InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS;

                break;
            }
        }

        view.setInputType(value);
    }

    protected void patchFontSize(Float fontSize) {
        float value = ((Integer) PixelHelper.dpToPx(fontSize != null ? fontSize : 16)).floatValue();

        view.setTextSize(TypedValue.COMPLEX_UNIT_PX, value);
    }

    protected void patchColor(Object color) {
        Integer tmpColor = ColorHelper.parseColor(color);

        view.setTextColor(tmpColor != null ? tmpColor : defaultColor);
    }

    private static class OnChangeData {
        String value;
    }

    protected static class Attributes extends Element.Attributes {
        String placeholder;
        Object placeholderColor;
        String value;
        String autoCapitalize;
        Boolean onChange;
    }
}

File: ios/Sources/Detonator/Element/TextAreaElement/TextAreaElement.swift
import UIKit

class TextAreaElement: Element, UIGestureRecognizerDelegate, UITextViewDelegate {
    private let defaultPlaceholderColor: UIColor = UIColor(white: 1, alpha: 0.75)
    
    override public func decodeAttributes(edge: Edge) -> TextAreaAttributes? {
        return super.decodeAttributes(edge: edge)
    }
    
    override public func createView() -> TextAreaView {
        let view = TextAreaView()
        
        view.delegate = self
        
        tapGestureRecognizer.delegate = self
        
        view.isScrollEnabled = true
        view.alwaysBounceHorizontal = false
        view.alwaysBounceVertical = true
        
        view.textContainer.lineBreakMode = .byWordWrapping
        view.textContainer.maximumNumberOfLines = 0
        
        return view
    }
    
    override public func patchView() {
        let view = view as! TextAreaView
        
        let attributes = attributes as! TextAreaAttributes
        let currentAttributes = currentAttributes as! TextAreaAttributes?
        
        let placeholder = attributes.placeholder
        let currentPlaceholder = currentAttributes?.placeholder
        
        let patchPlaceholderBool = forcePatch || placeholder != currentPlaceholder
        
        if patchPlaceholderBool {
            view.placeholder.text = placeholder
        }
        
        let placeholderColor = attributes.placeholderColor
        let currentPlaceholderColor = currentAttributes?.placeholderColor
        
        let patchPlaceholderColorBool = forcePatch || CompareHelper.compareStyleColors(placeholderColor, currentPlaceholderColor)
        
        if patchPlaceholderColorBool {
            view.placeholder.textColor = placeholderColor != nil ? ColorHelper.parseColor(color: placeholderColor!) : defaultPlaceholderColor
        }
        
        let value = attributes.value
        
        let patchValueBool = forcePatch
        
        if patchValueBool {
            view.text = value
        }
        
        let autoCapitalize = attributes.autoCapitalize
        let currentAutoCapitalize = currentAttributes?.autoCapitalize
        
        let patchAutoCapitalizeBool = forcePatch || autoCapitalize != currentAutoCapitalize
        
        if patchAutoCapitalizeBool {
            patchInputType(autoCapitalize: autoCapitalize)
        }
    }
    
    private func patchInputType(autoCapitalize: String?) {
        let view = view as! TextAreaView
        
        switch autoCapitalize {
        case "characters":
            view.autocapitalizationType = .allCharacters
            
            break
            
        case "words":
            view.autocapitalizationType = .words
            
            break
            
        case "sentences":
            view.autocapitalizationType = .sentences
            
            break
            
        case "none":
            view.autocapitalizationType = .none
            
            break
            
        default:
            view.autocapitalizationType = .sentences
            
            break
        }
        
        if view.isFirstResponder {
            view.resignFirstResponder()
            
            view.becomeFirstResponder()
        }
    }
    
    override func patchPadding(
        padding: Float?,
        paddingHorizontal: Float?,
        paddingVertical: Float?,
        paddingTop: Float?,
        paddingLeft: Float?,
        paddingBottom: Float?,
        paddingRight: Float?
    ) {
        let view = view as! TextAreaView
        
        let tmpPadding = padding ?? 0
        
        var tmpPaddingTop = tmpPadding
        var tmpPaddingLeft = tmpPadding
        var tmpPaddingBottom = tmpPadding
        var tmpPaddingRight = tmpPadding
        
        if paddingHorizontal != nil {
            tmpPaddingLeft = paddingHorizontal!
            tmpPaddingRight = paddingHorizontal!
        }
        
        if paddingVertical != nil {
            tmpPaddingTop = paddingVertical!
            tmpPaddingBottom = paddingVertical!
        }
        
        if paddingTop != nil {
            tmpPaddingTop = paddingTop!
        }
        
        if paddingLeft != nil {
            tmpPaddingLeft = paddingLeft!
        }
        
        if paddingBottom != nil {
            tmpPaddingBottom = paddingBottom!
        }
        
        if paddingRight != nil {
            tmpPaddingRight = paddingRight!
        }
        
        view.textContainerInset = UIEdgeInsets(
            top: CGFloat(tmpPaddingTop),
            left: CGFloat(tmpPaddingLeft),
            bottom: CGFloat(tmpPaddingBottom),
            right: CGFloat(tmpPaddingRight)
        )
        
        view.placeholder.layoutParams.padding = LayoutInsets(
            top: tmpPaddingTop,
            left: tmpPaddingLeft,
            bottom: tmpPaddingBottom,
            right: tmpPaddingRight
        )
    }
    
    override func patchFontSize(fontSize: Float?) {
        let view = view as! TextAreaView
        
        let value = CGFloat(fontSize ?? 16)
        
        view.font = UIFont.systemFont(ofSize: value)
        
        view.placeholder.font = view.font
    }
    
    override func patchColor(color: StyleColor?) {
        let view = view as! TextAreaView
        
        view.textColor = color != nil ? ColorHelper.parseColor(color: color!) : nil
    }
    
    func textViewDidChange(_ textView: UITextView) {
        let data = OnChangeData(value: textView.text ?? "")
        
        detonator.emitHandler(name: "onChange", edgeId: edge.id, data: data)
        
        detonator.performLayout()
    }
    
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
        return view.isFirstResponder
    }
    
    struct OnChangeData: Encodable {
        let value: String
    }
    
    class TextAreaAttributes: Attributes {
        var placeholder: String?
        var placeholderColor: StyleColor?
        var value: String?
        var autoCapitalize: String?
        var onChange: Bool?
        
        required init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            placeholder = try container.decodeIfPresent(String.self, forKey: .placeholder)
            placeholderColor = try container.decodeIfPresent(StyleColor.self, forKey: .placeholderColor)
            value = try container.decodeIfPresent(String.self, forKey: .value)
            autoCapitalize = try container.decodeIfPresent(String.self, forKey: .autoCapitalize)
            onChange = try container.decodeIfPresent(Bool.self, forKey: .onChange)
            
            try super.init(from: decoder)
        }
        
        private enum CodingKeys: String, CodingKey {
            case placeholder
            case placeholderColor
            case value
            case autoCapitalize
            case onChange
        }
    }
}
